### PR TITLE
Agent 3.45

### PR DESCRIPTION
**Pull request by moskyb**
**This PR:** Updates the agent docs for [v3.45](https://github.com/buildkite/agent/pull/2018), adding documentation for the new features and commands that it adds.

### CODE REVIEW

File: pages/agent/v3.md [hunk 1]
```diff
@@ -135,6 +135,37 @@ When the experiment is enabled, the agent will use different lock files from age
...
+Exposes a local API for the currently running job to introspect and mutate its state in the form of environment variables. This allows you to write scripts, hooks and plugins in languages other than bash, using them to interact with the agent.
+
+The API is exposed via a Unix Domain Socket, whose path is exposed to running jobs with the `BUILDKITE_AGENT_JOB_API_SOCKET` environment variable, and authenticated with a token exposed using the `BUILDKITE_AGENT_JOB_API_TOKEN` environment variable, using the `Bearer` HTTP Authorization scheme.
+
+The API exposes the following endpoints:
+
+- `GET /api/current-job/v0/env` - returns a JSON object of all environment variables for the current job
+- `PATCH /api/current-job/v0/env` - accepts a JSON object of environment variables to set for the current job
+- `DELETE /api/current-job/v0/env` - accepts a JSON array of environment variable names to unset for the current job
+
```
Comment (moskyb) [inline pages/agent/v3.md:161]: the formatting of this is getting mangled in render:
CleanShot 2023-03-17 at 15 38 30@2x

@mbelton-buildkite do you know why that might be?

Comment (mbelton-buildkite) [reply pages/agent/v3.md:161]: Hmm not sure what caused that. Are you still seeing it that way? The list looks okay when I open [the page on Render](https://bk-docs-pr-1954.onrender.com/docs/agent/v3):

Screenshot 2023-03-17 at 4 28 18 pm

Comment (moskyb) [reply pages/agent/v3.md:161]: oh, i think i might've made that screenshot after i added a newline before the list, but maybe before the render preview compiled. no worries!

File: pages/agent/v3/cli_env.md [hunk 2]
```diff
@@ -5,7 +5,22 @@ The Buildkite Agent's `env` subcommands provide the ability to inspect environme
 
 ## Printing env
 
+This command is used internally by the Buildkite Agent and isn't recommended for use in your builds.
```
Comment (mbelton-buildkite) [inline pages/agent/v3/cli_env.md:8]: Thoughts on just calling it "the agent"? I think the context is clear 🤔

```suggestion
This command is used internally by the agent and isn't recommended for use in your builds.
```

File: pages/agent/v3/cli_env.md [hunk 3]
```diff
@@ -5,7 +5,22 @@ The Buildkite Agent's `env` subcommands provide the ability to inspect environme
...
 
 <%= render 'agent/v3/help/env_dump' %>
 
+## Inspecting and modifying env from within a job
 
+Jobs can inspect and modify their own environment variables using the `env get`, `set`, and `unset` commands, so long as the job-api experiment is enabled. These commands provide an alternative to using shell commands to inspect and modify environment variables.
```
Comment (mbelton-buildkite) [inline pages/agent/v3/cli_env.md:14]: Style nitpicks:

```suggestion
Jobs can inspect and modify their environment variables using the `env get`, `set`, and `unset` commands, so long as the `job-api` experiment is enabled. These commands provide an alternative to using shell commands to inspect and modify environment variables.
```

File: pages/agent/v3/help/_bootstrap.md [hunk 4]
```diff
@@ -65,13 +65,14 @@ See https://buildkite.com/docs/agent/v3/hooks for more details.
...
-<tr id="git-submodule-clone-config"><th><code>--git-submodule-clone-config value</code> <a class="Docs__attribute__link" href="#git-submodule-clone-config">#</a></th><td><p>Comma separated key=value git config pairs applied before git submodule clone commands. For example, <code>update --init</code>. If the config is needed to be applied to all git commands, supply it in a global git config file for the system that the agent runs in instead.<br /><strong>Environment variable</strong>: <code>$BUILDKITE_GIT_SUBMODULE_CLONE_CONFIG</code></p></td></tr>
+<tr id="git-submodule-clone-config"><th><code>--git-submodule-clone-config value</code> <a class="Docs__attribute__link" href="#git-submodule-clone-config">#</a></th><td><p>Comma separated key=value git config pairs applied before git submodule clone commands. For example, `update --init`. If the config is needed to be applied to all git commands, supply it in a global git config file for the system that the agent runs in instead.<br /><strong>Environment variable</strong>: <code>$BUILDKITE_GIT_SUBMODULE_CLONE_CONFIG</code></p></td></tr>
```
Comment (mbelton-buildkite) [inline pages/agent/v3/help/_bootstrap.md:68]: What's the reason for this change? It looks like it shows the backticks rather than using the monospace code font.

File: pages/agent/v3/help/_bootstrap.md [hunk 5]
```diff
@@ -65,13 +65,14 @@ See https://buildkite.com/docs/agent/v3/hooks for more details.
...
 <tr id="bin-path"><th><code>--bin-path value</code> <a class="Docs__attribute__link" href="#bin-path">#</a></th><td><p>Directory where the buildkite-agent binary lives<br /><strong>Environment variable</strong>: <code>$BUILDKITE_BIN_PATH</code></p></td></tr>
 <tr id="build-path"><th><code>--build-path value</code> <a class="Docs__attribute__link" href="#build-path">#</a></th><td><p>Directory where builds will be created<br /><strong>Environment variable</strong>: <code>$BUILDKITE_BUILD_PATH</code></p></td></tr>
 <tr id="hooks-path"><th><code>--hooks-path value</code> <a class="Docs__attribute__link" href="#hooks-path">#</a></th><td><p>Directory where the hook scripts are found<br /><strong>Environment variable</strong>: <code>$BUILDKITE_HOOKS_PATH</code></p></td></tr>
+<tr id="sockets-path"><th><code>--sockets-path value</code> <a class="Docs__attribute__link" href="#sockets-path">#</a></th><td><p>Directory where the agent will place sockets (default: "/Users/ben/.buildkite-agent/sockets")<br /><strong>Environment variable</strong>: <code>$BUILDKITE_SOCKETS_PATH</code></p></td></tr>
```
Comment (mbelton-buildkite) [inline pages/agent/v3/help/_bootstrap.md:75]: "will" is another word that is often a smell:

```suggestion
<tr id="sockets-path"><th><code>--sockets-path value</code> <a class="Docs__attribute__link" href="#sockets-path">#</a></th><td><p>Directory where the agent places sockets (default: "/Users/ben/.buildkite-agent/sockets")<br /><strong>Environment variable</strong>: <code>$BUILDKITE_SOCKETS_PATH</code></p></td></tr>
```

File: pages/agent/v3/help/_env_dump.md [hunk 6]
```diff
@@ -13,9 +13,8 @@ script.
 -->
 
 ### Usage
-
-`buildkite-agent env dump [options]`
-
+buildkite-agent env dump [options]
+``
```
Comment (mbelton-buildkite) [inline pages/agent/v3/help/_env_dump.md:17]: This looks like a mistake:

Screenshot 2023-03-21 at 10 15 38 am

File: pages/agent/v3/help/_env_get.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+ | |__| | |__| | | |\  | |__| | | |    | |____| |__| || |_   | |
+ |_____/ \____/  |_| \_|\____/  |_|    |______|_____/_____|  |_|
+
+This file is auto-generated by scripts/update-agent-help.sh, please update the
+agent CLI help in https://github.com/buildkite/agent and run the generation
+script.
+
+-->
+
+### Usage
```
Comment (mbelton-buildkite) [inline pages/agent/v3/help/_env_get.md:15]: The heading levels for this don't look right. I'm guessing it's an issue with the script? If it's not easy to change that, perhaps we should change the structure of the page to make it work.

Screenshot 2023-03-21 at 10 17 29 am

File: pages/agent/v3/help/_env_get.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+`buildkite-agent env get [variables]`
+
+### Description
+Retrieves environment variables and their current values from the current job
```
Comment (mbelton-buildkite) [inline pages/agent/v3/help/_env_get.md:20]: I think we can avoid this repetition without losing the meaning:

```suggestion
Retrieves environment variables and their values from the current job
```

File: pages/agent/v3/help/_env_get.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+execution environment. 
+
+Note that this subcommand is only available from within the job executor with
+the ′job-api′ experiment enabled.
+
+Changes to the job environment only apply to the environments of subsequent
+phases of the job. However, ′env get′ can be used to inspect the changes made
```
Comment (mbelton-buildkite) [inline pages/agent/v3/help/_env_get.md:27]: Nitpick + use backticks for code format:

```suggestion
job phases. However, `env get` can be used to inspect the changes made
```

File: pages/agent/v3/help/_env_get.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+with ′env set′ and ′env unset′.
```
Comment (mbelton-buildkite) [inline pages/agent/v3/help/_env_get.md:28]: Use backticks for inline code format:

```suggestion
with `env set` and `env unset`.
```

File: pages/agent/v3/help/_env_get.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+Example (gets all variables in key=value format):
+
+    $ buildkite-agent env get
+ALPACA=Geronimo the Incredible
+BUILDKITE=true
+LLAMA=Kuzco
+...
```
Comment (mbelton-buildkite) [inline pages/agent/v3/help/_env_get.md:36]: Is this how the formatting needs to work in the CLI? For the docs, it would be nice to be something more like:

````
To get all the variables in `key=value` format, run:

```
buildkite-agent env get
```

The results will look like the following:

```
ALPACA=Geronimo the Incredible, BUILDKITE=true, LLAMA=Kuzco
```
````

File: pages/agent/v3/help/_env_set.md
Comment (mbelton-buildkite) [inline pages/agent/v3/help/_env_set.md:1]: Similar comments to the `env get` command

File: pages/agent/v3/help/_env_unset.md
Comment (mbelton-buildkite) [inline pages/agent/v3/help/_env_unset.md:1]: Similar comments to the `env get` command.

File: pages/agent/v3/help/_env_unset.md [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+script.
+
+-->
+
+### Usage
+
+`buildkite-agent env unset [variables]`
+
+### Description
+Un-sets environment variables in the current job execution environment.
```
Comment (mbelton-buildkite) [inline pages/agent/v3/help/_env_unset.md:20]: What do you think about removing the hyphen here?

```suggestion
Unsets environment variables in the current job execution environment. 
```

File: pages/agent/v3/help/_start.md [hunk 13]
```diff
@@ -67,6 +67,7 @@ The agent will run any jobs within a PTY (pseudo terminal) if available.
 <tr id="bootstrap-script"><th><code>--bootstrap-script value</code> <a class="Docs__attribute__link" href="#bootstrap-script">#</a></th><td><p>The command that is executed for bootstrapping a job, defaults to the bootstrap sub-command of this binary<br /><strong>Environment variable</strong>: <code>$BUILDKITE_BOOTSTRAP_SCRIPT_PATH</code></p></td></tr>
 <tr id="build-path"><th><code>--build-path value</code> <a class="Docs__attribute__link" href="#build-path">#</a></th><td><p>Path to where the builds will run from<br /><strong>Environment variable</strong>: <code>$BUILDKITE_BUILD_PATH</code></p></td></tr>
 <tr id="hooks-path"><th><code>--hooks-path value</code> <a class="Docs__attribute__link" href="#hooks-path">#</a></th><td><p>Directory where the hook scripts are found<br /><strong>Environment variable</strong>: <code>$BUILDKITE_HOOKS_PATH</code></p></td></tr>
+<tr id="sockets-path"><th><code>--sockets-path value</code> <a class="Docs__attribute__link" href="#sockets-path">#</a></th><td><p>Directory where the agent will place sockets (default: "/Users/ben/.buildkite-agent/sockets")<br /><strong>Environment variable</strong>: <code>$BUILDKITE_SOCKETS_PATH</code></p></td></tr>
```
Comment (mbelton-buildkite) [inline pages/agent/v3/help/_start.md:70]: As suggested earlier:

```suggestion
<tr id="sockets-path"><th><code>--sockets-path value</code> <a class="Docs__attribute__link" href="#sockets-path">#</a></th><td><p>Directory where the agent places sockets (default: "/Users/ben/.buildkite-agent/sockets")<br /><strong>Environment variable</strong>: <code>$BUILDKITE_SOCKETS_PATH</code></p></td></tr>
```

**Review by mbelton-buildkite — COMMENTED** [review]
A whole bunch of comments, but nothing too major and a bit of me just checking my understanding.

File: pages/agent/v3.md [hunk 14]
```diff
@@ -135,6 +135,37 @@ When the experiment is enabled, the agent will use different lock files from age
 >🛠 Experimental feature
 > To use flock file locks, set <code>experiment="flock-file-locks"</code> in your <a href="/docs/agent/v3/configuration#experiment"> agent configuration</a>.
 
+### Kubernetes exec
+
+Modifies `start` and `bootstrap` in such a way that they can run in separate Kubernetes containers in the same pod.
+
+Currently, this experiment is being used by [agent-stack-k8s](https://github.com/buildkite/agent-stack-k8s).
+
+This will result in errors unless orchestrated in a similar manner to that project. Please see the [README](https://github.com/buildkite/agent-stack-k8s/blob/main/README.md) of that repository for more details.
```
[the remark this answers is not in the document]
Comment (moskyb) [reply pages/agent/v3.md:144]: i think the main drive of this sentence is "if you just use the experiment, then nothing will happen and it'll be broken. you need to be using the agent-stack-k8s for it to work properly", but honestly i'm not 100% sure. @triarius do you have thoughts?

**Comment by triarius** [comment]
@moskyb I will take over the kuberenetes-exec comments and leave the job-api ones to you

File: pages/agent/v3/help/_bootstrap.md (hunk 4 above)
Comment (moskyb) [reply pages/agent/v3/help/_bootstrap.md:68]: i'm not 100% sure why this changed, but i suspect it has to do with [prime signs](https://github.com/buildkite/agent/blob/main/clicommand/prime-signs.md?plain=1) thing, which has historically been a bit janky and doesn't work that great

File: pages/agent/v3/help/_env_dump.md (hunk 6 above)
Comment (moskyb) [inline pages/agent/v3/help/_env_dump.md:17]:
```suggestion
`buildkite-agent env dump [options]`
```

File: pages/agent/v3/help/_env_unset.md (hunk 12 above)
Comment (moskyb) [reply pages/agent/v3/help/_env_unset.md]: lol: 
CleanShot 2023-03-21 at 14 24 28@2x

File: pages/agent/v3/help/_bootstrap.md [hunk 15]
```diff
@@ -65,13 +65,14 @@ See https://buildkite.com/docs/agent/v3/hooks for more details.
...
 <tr id="bin-path"><th><code>--bin-path value</code> <a class="Docs__attribute__link" href="#bin-path">#</a></th><td><p>Directory where the buildkite-agent binary lives<br /><strong>Environment variable</strong>: <code>$BUILDKITE_BIN_PATH</code></p></td></tr>
 <tr id="build-path"><th><code>--build-path value</code> <a class="Docs__attribute__link" href="#build-path">#</a></th><td><p>Directory where builds will be created<br /><strong>Environment variable</strong>: <code>$BUILDKITE_BUILD_PATH</code></p></td></tr>
 <tr id="hooks-path"><th><code>--hooks-path value</code> <a class="Docs__attribute__link" href="#hooks-path">#</a></th><td><p>Directory where the hook scripts are found<br /><strong>Environment variable</strong>: <code>$BUILDKITE_HOOKS_PATH</code></p></td></tr>
+<tr id="sockets-path"><th><code>--sockets-path value</code> <a class="Docs__attribute__link" href="#sockets-path">#</a></th><td><p>Directory where the agent places sockets (default: "/Users/ben/.buildkite-agent/sockets")<br /><strong>Environment variable</strong>: <code>$BUILDKITE_SOCKETS_PATH</code></p></td></tr>
```
Comment (triarius) [inline pages/agent/v3/help/_bootstrap.md:75]:
```suggestion
<tr id="sockets-path"><th><code>--sockets-path value</code> <a class="Docs__attribute__link" href="#sockets-path">#</a></th><td><p>Directory where the agent places sockets (default: "$HOME/.buildkite-agent/sockets")<br /><strong>Environment variable</strong>: <code>$BUILDKITE_SOCKETS_PATH</code></p></td></tr>
```

I think we should hard code the value to use `$HOME` in https://github.com/buildkite/agent/blob/ac3643bdd906ae422398094708cab09a0e324db2/clicommand/bootstrap.go#L276

Comment (triarius) [reply pages/agent/v3/help/_bootstrap.md:75]: And whatever it is on windows...

File: pages/agent/v3/help/_env_get.md (hunk 11 above)
Comment (moskyb) [reply pages/agent/v3/help/_env_get.md:36]: @mbelton-buildkite yeah. this all boils down to the process that we use to export CLI docs to the docs repo being super crufty, and not really how the cli docs are meant to be interpreted.

what i might do is kick this can down the road a bit by manually editing these files (in the docs repo) to be in the right format, leaving the next person (who will most likely also be me) to clean up the mess next time the docs get autogenerated.

it's probably also worth having a convo at some point about improving this process so that this whole feedback loop is quicker and easier.

File: pages/agent/v3.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (mbelton-buildkite) [reply pages/agent/v3.md:144]: Yip, I think that's clearer. Can you include that in the paragraph before? So it's:

> The [Buildkite Agent Stack for Kubernetes](https://github.com/buildkite/agent-stack-k8s) uses this experiment to run Buildkite steps as Kubernetes jobs. It is not expected to work outside of that context. See the [README](https://github.com/buildkite/agent-stack-k8s/blob/main/README.md) for more details.

File: pages/agent/v3/help/_bootstrap.md (hunk 4 above)
Comment (mbelton-buildkite) [reply pages/agent/v3/help/_bootstrap.md:68]: 😆 I really appreciate that you documented that. TIL: prime signs. 

I assume this is another quirk of the doc generation. I see that other fields have the same issue in their descriptions, but the example values display correctly 🫠

File: pages/agent/v3/help/_env_set.md [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+`buildkite-agent env set [variable]`
+
+### Description
+Sets environment variable values in the current job execution environment.  Existing variables will be overwritten.
+
+Note that this subcommand is only available from within the job executor with the `job-api` experiment enabled.
+
+Note that changes to the job environment variables only apply to subsequent phases of the job. To read the new values of variables from within the current phase, use `env get`.
+
+Note that Buildkite read-only variables cannot be overwritten.
```
Comment (mbelton-buildkite) [inline pages/agent/v3/help/_env_set.md:26]: There's a lot of _noting_ going on here. If possible, perhaps simplify to:

```suggestion
Unsets environment variables in the current job execution environment. Changes to the job environment variables only apply to subsequent phases of the job. This command cannot unset Buildkite read-only variables.

To read the new values of variables from within the current phase, use `env get`.

Note that this subcommand is only available from within the job executor with the job-api experiment enabled.
```

**Review by mbelton-buildkite — COMMENTED** [review]
A few more thoughts to consider, but I think this is basically there :)

File: pages/agent/v3.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (triarius) [reply pages/agent/v3.md:144]: Fixed in 7d0e1715

File: pages/agent/v3/cli_env.md [hunk 17]
```diff
@@ -5,7 +5,22 @@ The Buildkite Agent's `env` subcommands provide the ability to inspect environme
 
 ## Printing env
 
+This command is used internally by the agent and isn't recommended for use in your builds.
 
 <%= render 'agent/v3/help/env_dump' %>
 
+## Inspecting and modifying env from within a job
 
+Jobs can inspect and modify their environment variables using the `env get`, `set`, and `unset` commands, so long as the `job-api` experiment is enabled. These commands provide an alternative to using shell commands to inspect and modify environment variables.
```
[the remark this answers is not in the document]
Comment (triarius) [reply pages/agent/v3/cli_env.md:14]: I think the fact you need the job-api experiment is also repeated in the imported documents for get, set and unset. So I agree it makes sense to remove the warning from the 

`## Inspecting and modifying env from within a job`

Section and move it to the intro of this page. (along with everything else in that section).

**Comment by triarius** [comment]
2023-03-23T16:22:53,719480773+11:00

File: pages/agent/v3/help/_env_get.md (hunk 11 above)
Comment (triarius) [reply pages/agent/v3/help/_env_get.md]: To me, it makes sense to display the result on multiple lines because it's supposed to simulate what the output would look like on the terminal.

Whether we put the command that the user should type and the result of that command in separate sections is less clear. I think the best experience is if the display were like the screenshot below, but there was a "copy" button that copied only the command (i.e. the line that has the `$ `, but excluding the `$ `). That would require changing the script a lot, though.

See the screenshot here https://github.com/buildkite/docs/pull/1954#issuecomment-1480621935

**Comment by triarius** [comment]
2023-03-23T16:33:09,128870642+11:00

File: pages/agent/v3/cli_env.md [hunk 18]
```diff
@@ -5,7 +5,22 @@ The Buildkite Agent's `env` subcommands provide the ability to inspect environme
...
+
+## Getting a job's environment variables
+
```
[the remark this answers is not in the document]
Comment (triarius) [reply pages/agent/v3/cli_env.md:14]: This info will be in the `### Descrption` in the imported document. Having it above the `### Usage` heading seems like too much repetition.

See the screenshot: https://github.com/buildkite/docs/pull/1954#issuecomment-1480628220

File: pages/agent/v3/cli_env.md [hunk 19]
```diff
@@ -5,7 +5,22 @@ The Buildkite Agent's `env` subcommands provide the ability to inspect environme
...
+<%= render 'agent/v3/help/env_get' %>
+
+## Setting a job's environment variables
+
```
[the remark this answers is not in the document]
Comment (triarius) [reply pages/agent/v3/cli_env.md:18]: As above.

File: pages/agent/v3/cli_env.md [hunk 20]
```diff
@@ -5,7 +5,22 @@ The Buildkite Agent's `env` subcommands provide the ability to inspect environme
...
+<%= render 'agent/v3/help/env_set' %>
+
+## Removing environment variables from a job
+
```
[the remark this answers is not in the document]
Comment (triarius) [reply pages/agent/v3/cli_env.md:22]: As above.

File: pages/agent/v3/cli_env.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (mbelton-buildkite) [reply pages/agent/v3/cli_env.md:14]: Yeah, okay. Let's leave it without.

File: pages/agent/v3/help/_env_get.md (hunk 11 above)
Comment (mbelton-buildkite) [reply pages/agent/v3/help/_env_get.md]: Yeah, I think there are a bunch of workarounds needed to make this better, so I'm happy to leave it as-is. Manually updating it runs the risk of being overwritten with a future update.

File: pages/agent/v3/help/_env_set.md [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+
+Example:
```
Comment (mbelton-buildkite) [inline pages/agent/v3/help/_env_set.md:28]: This should be plural like the others:

```suggestion
Examples:
```

**Review by mbelton-buildkite — APPROVED** [review]
I've caught one last thing and there's my [comment about noting](https://github.com/buildkite/docs/pull/1954/files#r1145627824) to consider, BUT I think that's it. This one took quite a bit so thanks for getting it over the line. I learned a bunch 😊

File: pages/agent/v3/help/_env_set.md (hunk 21 above)
Comment (triarius) [reply pages/agent/v3/help/_env_set.md]: The `Example` in `buildkite-agent env dump` is an H3, so I made these H3 too.
